libevmasm/Assembly.cpp

```diff
@@ -298,6 +477,64 @@ Json::Value Assembly::assemblyJSON(map<string, unsigned> const& _sourceIndices,
 	return root;
 }
 
+shared_ptr<Assembly> Assembly::fromJSON(Json::Value const& _json, vector<string> const& _sourceList, int _level)
+{
+	solRequire(_json.isObject(), AssemblyImportException, "Supplied JSON is not an object.");
+	static set<string> const validMembers{".code", ".data", ".auxdata", "sourceList"};
+	for (auto const& attribute: _json.getMemberNames())
+		solRequire(validMembers.count(attribute), AssemblyImportException, "Unknown attribute '" + attribute + "'.");
+	solRequire(_json.isMember(".code"), AssemblyImportException, "Attribute '.code' does not exist.");
+	solRequire(_json[".code"].isArray(), AssemblyImportException, "Attribute '.code' is not an array.");
+
+	if (_level == 0)
+	{
+		solRequire(_json.isMember("sourceList"), AssemblyImportException, "Attribute 'sourceList' does not exist.");
+		solRequire(_json["sourceList"].isArray(), AssemblyImportException, "Attribute 'sourceList' is not an array.");
+	} else
+		solRequire(
+			!_json.isMember("sourceList"),
+			AssemblyImportException,
+			"Attribute 'sourceList' is only allowed in root JSON object."
+		);
+
+	shared_ptr<Assembly> result = make_shared<Assembly>(_level == 0, "");
```

Comment:
While reviewing https://github.com/ethereum/solidity/pull/12941#discussion_r1031688576 I realized that this is not the full condition for a creation object. If the object name ends with `_deployed`, it's a creation object as well.

Do objects even have names when compiled to assembly? If they do not, then this feature might not be able to properly handle Yul code compiled to EVM if the user is not careful about adding the `_deployed` suffix to all non-top-level objects.

Reply:
Good catch! As of now it is not possible to export assembly json from yul, so the resulting data objects will not contain any names. I added now an additional check for the `.data` keys: all keys need to be parseable integers. 

However, we could later ensure that there is only one `_deployed` object, where that object will just become the first object. But for this we need to be able to export yul to assembly json.

Reply:
> I added now an additional check for the .data keys: all keys need to be parseable integers.

Ok, I just saw that some `.data` keys can indeed be real strings. So I removed the need that they need to be parseable integers. I think the best is really just to take care of that during the export of yul to assembly json. The `_deployed` just need to be the first within the `.data` section.